components/vocabularies/api/src/main/java/org/phenotips/vocabulary/internal/solr/DefaultSolrVocabularyResourceManager.java

```diff
@@ -59,16 +73,84 @@
 
+    @Inject
+    private Environment environment;
+
     @Override
     public void initialize(String vocabularyName) throws InitializationException
     {
+        // Get jars home path to where the jar resources are stored
+        Class<DefaultSolrVocabularyResourceManager> clazz = DefaultSolrVocabularyResourceManager.class;
+        String jarPath = clazz.getProtectionDomain().getCodeSource().getLocation().getPath();
+        String jarsHome = jarPath.substring(0, jarPath.lastIndexOf('/'));
+        // Get data Solr home path
+        File solrHome = new File(this.environment.getPermanentDirectory().getAbsolutePath(), "solr");
+        File dest = solrHome;
+
+        CoreContainer container = this.cores.getContainer();
+
+        // Check if the core doesn't exist already
+        if (container.getCore(vocabularyName) != null) {
+            return;
+        }
+
         try {
-            this.core = new EmbeddedSolrServer(this.cores.getContainer(), vocabularyName);
+            File dir = new File(jarsHome);
+            File[] directoryListing = dir.listFiles();
+
+            if (directoryListing == null) {
+                return;
+            }
+
+            for (File file : directoryListing) {
+                JarFile jar = new JarFile(file);
+                String jarName = jar.getName().substring(jarsHome.length());
+                if ("jar".equals(FilenameUtils.getExtension(file.toPath().toString()))
+                    && jarName.startsWith("vocabulary-" + vocabularyName)) {
+                    copyConfigsFromJar(jar, vocabularyName, dest);
+                }
+            }
+
+            CoreDescriptor dcore =
+                new CoreDescriptor(container, vocabularyName, solrHome.toPath().resolve(vocabularyName).toString());
+
+            this.core = new EmbeddedSolrServer(container, vocabularyName);
+
+            this.score = container.create(dcore);
+
             this.cache = this.cacheFactory.createNewLocalCache(new CacheConfiguration());
         } catch (RuntimeException ex) {
             throw new InitializationException("Invalid Solr core: " + ex.getMessage());
         } catch (final CacheException ex) {
             throw new InitializationException("Cannot create cache: " + ex.getMessage());
+        } catch (IOException ex) {
+            throw new InitializationException("Invalid Solr resource: " + ex.getMessage());
+        }
+    }
+
+    /**
+     * Copy configuration files from a jar file specified by jar to a destination folder on the filesystem dest.
+     *
+     * @param jar the jar file
+     * @param vocabularyName the name of the vocabulary being managed
+     * @param dest the destination
+     * @throws IOException
+     */
+    private static void copyConfigsFromJar(JarFile jar, String vocabularyName, File dest) throws IOException
+    {
+        // return if vocabulary directory already exists
+        if (Files.isDirectory(dest.toPath().resolve(vocabularyName))) {
```

Comment:
The assumption is that if the core files have already been extracted during the previous startup, then it's not necessary to re-copy them over.

However, given that it's not that costly to copy a few files, we should copy them each time. This also ensures that after an upgrade, the new version of the files will be used.

So yes, remove that condition.